File: src/main/java/rs/kunpero/fatpak/config/SecurityConfig.java
package rs.kunpero.fatpak.config;

import com.slack.api.app_backend.events.servlet.SlackSignatureVerifier;
import org.springframework.boot.web.servlet.FilterRegistrationBean;
import org.springframework.context.annotation.Bean;
import org.springframework.core.Ordered;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import rs.kunpero.fatpak.config.filter.SlackRequestVerifierFilter;

@EnableWebSecurity
public class SecurityConfig extends WebSecurityConfigurerAdapter {
    @Override
    public void configure(HttpSecurity http) throws Exception {
        http
                .authorizeRequests()
                .mvcMatchers("/feed/**")
                .permitAll()
                .and()
                .requiresChannel()
                .requestMatchers(r -> r.getHeader("X-Forwarded-Proto") != null)
                .requiresSecure()
                .and().csrf().disable();
    }

    @Bean
    public SlackRequestVerifierFilter slackRequestVerifierFilter(SlackSignatureVerifier slackSignatureVerifier) {
        return new SlackRequestVerifierFilter(slackSignatureVerifier);
    }

    @Bean
    public FilterRegistrationBean slackRequestVerifierFilterRegistration(SlackRequestVerifierFilter filter) {
        FilterRegistrationBean registration = new FilterRegistrationBean();
        registration.setFilter(filter);
        registration.addUrlPatterns("/feed/*");
        registration.setOrder(Ordered.HIGHEST_PRECEDENCE);
        return registration;
    }
}

File: src/main/kotlin/rs/kunpero/fatpak/config/SecurityConfig.kt
package rs.kunpero.fatpak.config

import com.slack.api.app_backend.events.servlet.SlackSignatureVerifier
import org.springframework.boot.web.servlet.FilterRegistrationBean
import org.springframework.context.annotation.Bean
import org.springframework.core.Ordered
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.web.util.matcher.RequestMatcher
import rs.kunpero.fatpak.config.filter.SlackRequestVerifierFilter
import javax.servlet.http.HttpServletRequest

@EnableWebSecurity
class SecurityConfig : WebSecurityConfigurerAdapter() {
    override fun configure(http: HttpSecurity) {
        http
            .authorizeRequests()
            .mvcMatchers("/feed/**")
            .permitAll()
            .and()
            .requiresChannel()
            .requestMatchers(RequestMatcher { r: HttpServletRequest ->
                r.getHeader(
                    "X-Forwarded-Proto"
                ) != null
            })
            .requiresSecure()
            .and().csrf().disable()
    }

    @Bean
    fun slackRequestVerifierFilter(slackSignatureVerifier: SlackSignatureVerifier): SlackRequestVerifierFilter {
        return SlackRequestVerifierFilter(slackSignatureVerifier)
    }

    @Bean
    fun slackRequestVerifierFilterRegistration(filter: SlackRequestVerifierFilter): FilterRegistrationBean<SlackRequestVerifierFilter> {
        val registration: FilterRegistrationBean<SlackRequestVerifierFilter> = FilterRegistrationBean()
        registration.filter = filter
        registration.addUrlPatterns("/feed/*")
        registration.order = Ordered.HIGHEST_PRECEDENCE
        return registration
    }
}